Orange/widgets/visualize/owmosaic.py

```diff
@@ -438,34 +440,29 @@ def set_data(self, data):
             self.init_combos(None)
             return
 
-        self.color_model.set_domain(self.data.domain)
         self.init_combos(self.data)
 
         self.openContext(self.data)
 
-        # if we first received subset we now call setSubsetData to process it
-        if self.unprocessed_subset_data:
-            self.set_subset_data(self.unprocessed_subset_data)
-            self.unprocessed_subset_data = None
-
-        self.set_color_data()
-
     @Inputs.data_subset
     def set_subset_data(self, data):
-        self.Warning.incompatible_subset.clear()
-        if self.data is None:
-            self.unprocessed_subset_data = data
-            return
-        try:
-            self.subset_data = data.transform(self.data.domain)
-        except:
-            self.subset_data = None
-            self.Warning.incompatible_subset(shown=data is not None)
+        self.subset_data = data
 
     # this is called by widget after setData and setSubsetData are called.
     # this way the graph is updated only once
     def handleNewSignals(self):
-        self.reset_graph()
+        self.Warning.incompatible_subset.clear()
+        self.subset_indices = indices = None
+        if self.data is not None and self.subset_data:
+            transformed = self.subset_data.transform(self.data.domain)
+            if np.all(np.isnan(transformed.X)) and np.all(np.isnan(transformed.Y)):
+                self.Warning.incompatible_subset()
+            else:
+                indices = {e.id for e in transformed}
+                self.subset_indices = [ex.id in indices for ex in self.data]
+
+        self.clear_selection()
```

Comment:
This is new. Is it necessary -- and desired?

Selection is already cleared upon receiving new data. This would also clear it upon receiving a subset, which is unnecessary and undesired, I guess. Have I overlooked something?